File: program_stack/algorithms/strings/trie/cpp/trie.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

struct Node {
    std::unordered_map<char, std::unique_ptr<Node>> next;
    bool end = false;
};

class Trie {
public:
    void insert(const std::string& word) {
        Node* cur = &root;
        for (char c : word) {
            auto& slot = cur->next[c];
            if (!slot) slot = std::make_unique<Node>();
            cur = slot.get();
        }
        cur->end = true;
    }

    bool search(const std::string& word) const {
        const Node* cur = walk(word);
        return cur && cur->end;
    }

    bool startsWith(const std::string& prefix) const {
        return walk(prefix) != nullptr;
    }

private:
    Node root;

    const Node* walk(const std::string& s) const {
        const Node* cur = &root;
        for (char c : s) {
            auto it = cur->next.find(c);
            if (it == cur->next.end()) return nullptr;
            cur = it->second.get();
        }
        return cur;
    }
};

int main() {
    Trie t;
    t.insert("apple");
    if (!t.search("apple") || t.search("app") || !t.startsWith("app")) {
        std::cerr << "apple/app checks failed\n";
        return 1;
    }
    t.insert("app");
    if (!t.search("app") || !t.startsWith("ap") || t.search("appl")) {
        std::cerr << "after insert app failed\n";
        return 1;
    }
    if (t.search("") || t.startsWith("b")) {
        std::cerr << "edge checks failed\n";
        return 1;
    }
    std::cout << "trie: ok\n";
    return 0;
}

File: program_stack/algorithms/strings/trie/java/Trie.java
import java.util.HashMap;
import java.util.Map;

public class Trie {
    private static class Node {
        Map<Character, Node> next = new HashMap<>();
        boolean end;
    }

    private final Node root = new Node();

    public void insert(String word) {
        Node cur = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            cur = cur.next.computeIfAbsent(c, k -> new Node());
        }
        cur.end = true;
    }

    public boolean search(String word) {
        Node cur = walk(word);
        return cur != null && cur.end;
    }

    public boolean startsWith(String prefix) {
        return walk(prefix) != null;
    }

    private Node walk(String s) {
        Node cur = root;
        for (int i = 0; i < s.length(); i++) {
            cur = cur.next.get(s.charAt(i));
            if (cur == null) return null;
        }
        return cur;
    }

    public static void main(String[] args) {
        Trie t = new Trie();
        t.insert("apple");
        if (!t.search("apple") || t.search("app") || !t.startsWith("app")) {
            System.err.println("apple/app checks failed");
            System.exit(1);
        }
        t.insert("app");
        if (!t.search("app") || !t.startsWith("ap") || t.search("appl")) {
            System.err.println("after insert app failed");
            System.exit(1);
        }
        if (t.search("") || t.startsWith("b")) {
            System.err.println("edge checks failed");
            System.exit(1);
        }
        System.out.println("trie: ok");
    }
}
